src/rust/cryptography-x509-validation/src/policy/extension.rs

```diff
@@ -0,0 +1,366 @@
+        Self {
+            oid,
+            validator: ExtensionValidator::NotPresent,
+        }
+    }
+
+    pub(crate) fn present(
+        oid: ObjectIdentifier,
+        criticality: Criticality,
+        validator: Option<PresentExtensionValidatorCallback<B>>,
+    ) -> Self {
+        Self {
+            oid,
+            validator: ExtensionValidator::Present {
+                criticality,
+                validator,
+            },
+        }
+    }
+
+    pub(crate) fn maybe_present(
+        oid: ObjectIdentifier,
+        criticality: Criticality,
+        validator: Option<MaybeExtensionValidatorCallback<B>>,
+    ) -> Self {
+        Self {
+            oid,
+            validator: ExtensionValidator::MaybePresent {
+                criticality,
+                validator,
+            },
+        }
+    }
+
+    pub(crate) fn permits(
+        &self,
+        policy: &Policy<'_, B>,
+        cert: &Certificate<'_>,
+        extensions: &Extensions<'_>,
+    ) -> Result<(), PolicyError> {
+        match (&self.validator, extensions.get_extension(&self.oid)) {
+            // Extension MUST NOT be present and isn't; OK.
+            (ExtensionValidator::NotPresent, None) => Ok(()),
+            // Extension MUST NOT be present but is; NOT OK.
+            (ExtensionValidator::NotPresent, Some(_)) => Err(PolicyError::Other(
+                "EE certificate contains prohibited extension",
+            )),
+            // Extension MUST be present but is not; NOT OK.
+            (ExtensionValidator::Present { .. }, None) => Err(PolicyError::Other(
+                "EE certificate is missing required extension",
+            )),
+            // Extension MUST be present and is; check it.
+            (
+                ExtensionValidator::Present {
+                    criticality,
+                    validator,
+                },
+                Some(extn),
+            ) => {
+                if !criticality.permits(extn.critical) {
+                    return Err(PolicyError::Other(
+                        "EE certificate extension has incorrect criticality",
+                    ));
+                }
+
+                // If a custom validator is supplied, apply it.
+                validator.map_or(Ok(()), |v| v(policy, cert, &extn))
+            }
+            // Extension MAY be present.
+            (
+                ExtensionValidator::MaybePresent {
+                    criticality,
+                    validator,
+                },
+                extn,
+            ) => {
+                // If the extension is present, apply our criticality check.
+                if extn
+                    .as_ref()
+                    .map_or(false, |extn| !criticality.permits(extn.critical))
+                {
+                    return Err(PolicyError::Other(
+                        "EE certificate extension has incorrect criticality",
+                    ));
+                }
```

Comment:
(From CI results) looks like this branch needs a unit test.

Reply:
[1ba15dd](https://github.com/pyca/cryptography/pull/9781/commits/1ba15ddca6d1d43478a3133d13b1661322f4cf10)